### ISSUE TITLE
PRAISE - ERC20.approve() doesn't approve to 0 first for tokens like USDT

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-dodo-judging#35]
PRAISE

medium

# ERC20.approve() doesn't approve to 0 first for tokens like USDT

## Summary
first reduce  allowance to 0 first for tokens like USDT
## Vulnerability Detail
There are numerous instances where the IERC20.approve() function is called only once without setting the allowance to zero. Some tokens, like USDT, require first reducing the address' allowance to zero by calling approve(_spender, 0). Transactions will revert when using tokens like USDT (see the approve() function requirement below [at line 201-203](https://etherscan.io/address/0xdac17f958d2ee523a2206206994597c13d831ec7#code)).
```solidity
function approve(address _spender, uint _value) public onlyPayloadSize(2 * 32) {

        **// To change the approve amount you first have to reduce the addresses`
        **//  allowance to zero by calling `approve(_spender, 0)` if it is not
        **//  already 0 to mitigate the race condition described here:
        //  https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
        require(!((_value != 0) && (allowed[msg.sender][_spender] != 0)));

        allowed[msg.sender][_spender] = _value;
        Approval(msg.sender, _spender, _value);
    }

```
## Impact
Transactions will revert when using tokens like USDT 

## Code Snippet
https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Pool/D3Funding.sol#L22

https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Pool/D3Funding.sol#L52

https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Pool/D3Funding.sol#L66

## Tool used

Manual Review

## Recommendation
 To change the approve amount you first have to reduce the addresses allowance to zero by calling `approve(_spender, 0)`

Duplicate of #203
